### Assignment 1: Feature Engineering 

<mark>Ethan Wen - 1/27/2022 - Professor Mishra</mark>

From Kaggle and the assignment posted, there are a **set of attributes (79) describing houses in Ames, Iowa**. We will try to **predict the price of the house** based on these features using regression techniques. 

We can find the dataset located here (training split of the original dataset): [Click Here](https://www.kaggle.com/competitions/house-prices-advanced-regression-techniques/data?select=train.csv)

We can find the descriptions of the columns right [here](https://www.kaggle.com/competitions/house-prices-advanced-regression-techniques/data?select=data_description.txt)

The training dataset and the descriptions of the column are named train.csv and data_description.txt respectively. 

### a. Load the dataset as a dataframe using pandas library. Print a few columns using head() function 

In [1]:
import pandas as pd
house_predict_df = pd.read_csv("train.csv")
house_predict_df.head()

Id  MSSubClass MSZoning  LotFrontage  LotArea Street Alley LotShape  \
0   1          60       RL         65.0     8450   Pave   NaN      Reg   
1   2          20       RL         80.0     9600   Pave   NaN      Reg   
2   3          60       RL         68.0    11250   Pave   NaN      IR1   
3   4          70       RL         60.0     9550   Pave   NaN      IR1   
4   5          60       RL         84.0    14260   Pave   NaN      IR1   

  LandContour Utilities  ... PoolArea PoolQC Fence MiscFeature MiscVal MoSold  \
0         Lvl    AllPub  ...        0    NaN   NaN         NaN       0      2   
1         Lvl    AllPub  ...        0    NaN   NaN         NaN       0      5   
2         Lvl    AllPub  ...        0    NaN   NaN         NaN       0      9   
3         Lvl    AllPub  ...        0    NaN   NaN         NaN       0      2   
4         Lvl    AllPub  ...        0    NaN   NaN         NaN       0     12   

  YrSold  SaleType  SaleCondition  SalePrice  
0   2008        WD         Normal     208500  
1   2007        WD         Normal     181500  
2   2008        WD         Normal     223500  
3   2006        WD        Abnorml     140000  
4   2008        WD         Normal     250000  

[5 rows x 81 columns]

There are 81 Columns instead of the proposed 79 explantory features but I think that those 2 are probably from the "Id" column and something else (maybe sale price since we are supposed to predict that). 

In [2]:
print(house_predict_df.columns)

Index(['Id', 'MSSubClass', 'MSZoning', 'LotFrontage', 'LotArea', 'Street',
       'Alley', 'LotShape', 'LandContour', 'Utilities', 'LotConfig',
       'LandSlope', 'Neighborhood', 'Condition1', 'Condition2', 'BldgType',
       'HouseStyle', 'OverallQual', 'OverallCond', 'YearBuilt', 'YearRemodAdd',
       'RoofStyle', 'RoofMatl', 'Exterior1st', 'Exterior2nd', 'MasVnrType',
       'MasVnrArea', 'ExterQual', 'ExterCond', 'Foundation', 'BsmtQual',
       'BsmtCond', 'BsmtExposure', 'BsmtFinType1', 'BsmtFinSF1',
       'BsmtFinType2', 'BsmtFinSF2', 'BsmtUnfSF', 'TotalBsmtSF', 'Heating',
       'HeatingQC', 'CentralAir', 'Electrical', '1stFlrSF', '2ndFlrSF',
       'LowQualFinSF', 'GrLivArea', 'BsmtFullBath', 'BsmtHalfBath', 'FullBath',
       'HalfBath', 'BedroomAbvGr', 'KitchenAbvGr', 'KitchenQual',
       'TotRmsAbvGrd', 'Functional', 'Fireplaces', 'FireplaceQu', 'GarageType',
       'GarageYrBlt', 'GarageFinish', 'GarageCars', 'GarageArea', 'GarageQual',
       'GarageCond', 'PavedDrive

### (b) Perform exploratory data analysis

Checking duplicates like in the imdb_dataframe lab from last week

In [3]:
house_predict_df.duplicated().sum()

0

Apparently there are no duplicates in this dataset. 

In [4]:
house_predict_df.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 1460 entries, 0 to 1459
Data columns (total 81 columns):
 #   Column         Non-Null Count  Dtype  
---  ------         --------------  -----  
 0   Id             1460 non-null   int64  
 1   MSSubClass     1460 non-null   int64  
 2   MSZoning       1460 non-null   object 
 3   LotFrontage    1201 non-null   float64
 4   LotArea        1460 non-null   int64  
 5   Street         1460 non-null   object 
 6   Alley          91 non-null     object 
 7   LotShape       1460 non-null   object 
 8   LandContour    1460 non-null   object 
 9   Utilities      1460 non-null   object 
 10  LotConfig      1460 non-null   object 
 11  LandSlope      1460 non-null   object 
 12  Neighborhood   1460 non-null   object 
 13  Condition1     1460 non-null   object 
 14  Condition2     1460 non-null   object 
 15  BldgType       1460 non-null   object 
 16  HouseStyle     1460 non-null   object 
 17  OverallQual    1460 non-null   int64  
 18  OverallC

It seems like there are some null values in some columns although most of them are non-null. You can also see the distribution of the different types of values with 38 of them being numerical (3 of them are floats and 35 of them are integers) and the remaining 43 values being an object(most likely a string). 

In [5]:
house_predict_df.count()

Id               1460
MSSubClass       1460
MSZoning         1460
LotFrontage      1201
LotArea          1460
                 ... 
MoSold           1460
YrSold           1460
SaleType         1460
SaleCondition    1460
SalePrice        1460
Length: 81, dtype: int64

The number of rows left are the same since there are no duplicated rows - should be 81 columns and 1460 rows. 

### (c) Feature and Label Definition

From the assignment, we choose these columns: Lotshape, LotConfig, Neighborhood, OverallQual, OverallCond, YearRemodAdd, GrLivArea, MoSold, and SalePrice.

In [6]:
temp = house_predict_df[["LotShape","LotConfig","Neighborhood","OverallQual", "OverallCond", "YearRemodAdd", "GrLivArea", "MoSold", "SalePrice"]] 

In [7]:
temp.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 1460 entries, 0 to 1459
Data columns (total 9 columns):
 #   Column        Non-Null Count  Dtype 
---  ------        --------------  ----- 
 0   LotShape      1460 non-null   object
 1   LotConfig     1460 non-null   object
 2   Neighborhood  1460 non-null   object
 3   OverallQual   1460 non-null   int64 
 4   OverallCond   1460 non-null   int64 
 5   YearRemodAdd  1460 non-null   int64 
 6   GrLivArea     1460 non-null   int64 
 7   MoSold        1460 non-null   int64 
 8   SalePrice     1460 non-null   int64 
dtypes: int64(6), object(3)
memory usage: 102.8+ KB


From here, we can see that all of the rows have no null values. Just in case, we can check...

In [8]:
temp = temp.dropna()
temp.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 1460 entries, 0 to 1459
Data columns (total 9 columns):
 #   Column        Non-Null Count  Dtype 
---  ------        --------------  ----- 
 0   LotShape      1460 non-null   object
 1   LotConfig     1460 non-null   object
 2   Neighborhood  1460 non-null   object
 3   OverallQual   1460 non-null   int64 
 4   OverallCond   1460 non-null   int64 
 5   YearRemodAdd  1460 non-null   int64 
 6   GrLivArea     1460 non-null   int64 
 7   MoSold        1460 non-null   int64 
 8   SalePrice     1460 non-null   int64 
dtypes: int64(6), object(3)
memory usage: 102.8+ KB


From here, we seperate the <mark> temp </mark> dataframe into two other dataframes/labels - <mark> housing_features_df and sales_price_df </mark>. 

In [9]:
sales_price_df = temp.pop('SalePrice')
sales_price_df.head()

0    208500
1    181500
2    223500
3    140000
4    250000
Name: SalePrice, dtype: int64

We can see that the values of SalePrice are numerical based on the saleprice of the house in dollars. 

In [10]:
housing_features_df = temp
housing_features_df = housing_features_df.dropna()
housing_features_df.head()


LotShape LotConfig Neighborhood  OverallQual  OverallCond  YearRemodAdd  \
0      Reg    Inside      CollgCr            7            5          2003   
1      Reg       FR2      Veenker            6            8          1976   
2      IR1    Inside      CollgCr            7            5          2002   
3      IR1    Corner      Crawfor            7            5          1970   
4      IR1       FR2      NoRidge            8            5          2000   

   GrLivArea  MoSold  
0       1710       2  
1       1262       5  
2       1786       9  
3       1717       2  
4       2198      12

From data_description.txt, we are given the values of each feature - most of them are categorical. 

In [11]:
housing_features_df["LotShape"].unique()

array(['Reg', 'IR1', 'IR2', 'IR3'], dtype=object)

The LotShape feature is **categorical ordinal** since there is some type of ordering (hierarchy). The following values are used for **LotShape**: Regular, Slighty Irregular, Moderately Irregular, Irregular. 

In [12]:
housing_features_df["LotConfig"].unique()

array(['Inside', 'FR2', 'Corner', 'CulDSac', 'FR3'], dtype=object)

From what I can tell, the LotConfig feature seems to be **categorical nominal** since it seems to be just different configurations. However, someone with more experience could probably determine if it was ordinal or not. The following values are used for LotConfig: Inside lot, corner lot, cul-de-sac, frontage on 2 sides of property, and frontage on 3 sides of property. 

In [13]:
housing_features_df["Neighborhood"].unique()

array(['CollgCr', 'Veenker', 'Crawfor', 'NoRidge', 'Mitchel', 'Somerst',
       'NWAmes', 'OldTown', 'BrkSide', 'Sawyer', 'NridgHt', 'NAmes',
       'SawyerW', 'IDOTRR', 'MeadowV', 'Edwards', 'Timber', 'Gilbert',
       'StoneBr', 'ClearCr', 'NPkVill', 'Blmngtn', 'BrDale', 'SWISU',
       'Blueste'], dtype=object)

The Neighborhood feature is **categorical nominal** since its just the names of neighborhoods which don't have any order/hierarchy. 

In [14]:
housing_features_df["OverallQual"].unique()

array([ 7,  6,  8,  5,  9,  4, 10,  3,  1,  2])

The OverallQual feature stands for overall quality, the numerical rating of the overall material and finish of the house. Even though the values are numerical, its a **categorical ordinal** since its on a scale from 1 to 10 with 1 being "Very Poor" and 10 being "Very Excellent".  

In [15]:
housing_features_df["OverallCond"].unique()

array([5, 8, 6, 7, 4, 2, 3, 9, 1])

Exactly like OverallQual, OverallCond feature stands for overall condition, the numerical rating of the overall condition of the house. Even though the values are numerical, its a **categorical ordinal** since its on a scale from 1 to 10 with 1 being "Very Poor" and 10 being "Very Excellent".  

In [16]:
housing_features_df["YearRemodAdd"].unique()

array([2003, 1976, 2002, 1970, 2000, 1995, 2005, 1973, 1950, 1965, 2006,
       1962, 2007, 1960, 2001, 1967, 2004, 2008, 1997, 1959, 1990, 1955,
       1983, 1980, 1966, 1963, 1987, 1964, 1972, 1996, 1998, 1989, 1953,
       1956, 1968, 1981, 1992, 2009, 1982, 1961, 1993, 1999, 1985, 1979,
       1977, 1969, 1958, 1991, 1971, 1952, 1975, 2010, 1984, 1986, 1994,
       1988, 1954, 1957, 1951, 1978, 1974])

The YearRemodAdd feature stands for the year of the date that the house was remodeled or the original construction date. It seems to be a **numerical value**. 

In [17]:
housing_features_df["GrLivArea"].unique()

array([1710, 1262, 1786, 1717, 2198, 1362, 1694, 2090, 1774, 1077, 1040,
       2324,  912, 1494, 1253,  854, 1004, 1296, 1114, 1339, 2376, 1108,
       1795, 1060, 1600,  900, 1704,  520, 1317, 1228, 1234, 1700, 1561,
       2452, 1097, 1297, 1057, 1152, 1324, 1328,  884,  938, 1150, 1752,
       2149, 1656, 1452,  955, 1470, 1176,  816, 1842, 1360, 1425, 1739,
       1720, 2945,  780, 1158, 1111, 1370, 2034, 2473, 2207, 1479,  747,
       2287, 2223,  845, 1718, 1086, 1605,  988,  952, 1285, 1768, 1230,
       2142, 1337, 1563, 1065, 1474, 2417, 1560, 1224, 1526,  990, 1235,
        964, 2291, 1588,  960,  835, 1225, 1610, 1732, 1535, 1226, 1818,
       1992, 1047,  789, 1517, 1844, 1855, 1430, 2696, 2259, 2320, 1458,
       1092, 1125, 3222, 1456, 1123, 1080, 1199, 1586,  754,  958,  840,
       1348, 1053, 2157, 2054, 1327, 1721, 1682, 1214, 1959, 1852, 1764,
        864, 1734, 1385, 1501, 1728, 1709,  875, 2035, 1344,  969, 1993,
       1252, 1200, 1096, 1968, 1947, 2462, 1232, 26

The GrLivArea feature stands for Above grade (ground) living area square feet. It's a **numerical value** since its based on square footage. 

In [18]:
housing_features_df["MoSold"].unique()

array([ 2,  5,  9, 12, 10,  8, 11,  4,  1,  7,  3,  6])

The MoSold feature stands for the month sold. It's a **categorical ordinal** value since months are ordered accordingly from 1 to 12 in a hierarchy.

In the dataframe, LotConfig and Neighborhood are features that are **categorical nominal** since they don't see to have any order or hierarchy. LotShape, OverallQual, OverallCond, and MoSold are all features that are **categorical ordinal** since they seem to be related in a hierarchial way. YearRemodAdd and GrLivArea are both features that are **numerical values**.  

### (d) Feature Transformation

For this part, we should take all the **categorical nominal** features and turn them into **one-hot encodings** (I'm not sure if we should do all categorical features so I'll do both to see how it works). So we should start with features LotConfig and Neighborhood. 

In [19]:
lot_config_list = housing_features_df["LotConfig"].to_list()
print(lot_config_list)

['Inside', 'FR2', 'Inside', 'Corner', 'FR2', 'Inside', 'Inside', 'Corner', 'Inside', 'Corner', 'Inside', 'Inside', 'Inside', 'Inside', 'Corner', 'Corner', 'CulDSac', 'Inside', 'Inside', 'Inside', 'Corner', 'Inside', 'Inside', 'Inside', 'Inside', 'Corner', 'Corner', 'Inside', 'CulDSac', 'Inside', 'Inside', 'CulDSac', 'Corner', 'Inside', 'Inside', 'Inside', 'Corner', 'Inside', 'Inside', 'Inside', 'Inside', 'Inside', 'CulDSac', 'CulDSac', 'Inside', 'Inside', 'CulDSac', 'Inside', 'Inside', 'Inside', 'Corner', 'Inside', 'Corner', 'Inside', 'Inside', 'Inside', 'Inside', 'Corner', 'CulDSac', 'Inside', 'Inside', 'Inside', 'Inside', 'Inside', 'Inside', 'Inside', 'Inside', 'Inside', 'Corner', 'Corner', 'Inside', 'Corner', 'Corner', 'Inside', 'Corner', 'Inside', 'Inside', 'Inside', 'Inside', 'Corner', 'Corner', 'FR2', 'Inside', 'Inside', 'Inside', 'Corner', 'Inside', 'Corner', 'Corner', 'Inside', 'Inside', 'Inside', 'Inside', 'Corner', 'Inside', 'Corner', 'Inside', 'Inside', 'Corner', 'Inside', '

Scikit responds to multiple features - list of lists (2D)

In [20]:
lot_config_list_of_lists = []

for i in lot_config_list: 
    lot_config_list_of_lists.append([i])

print(lot_config_list_of_lists[:5])

[['Inside'], ['FR2'], ['Inside'], ['Corner'], ['FR2']]


In [21]:
from sklearn.preprocessing import OneHotEncoder

lot_config_encoder = OneHotEncoder()

lot_config_encoder.fit(lot_config_list_of_lists)

print(f"Unique vocabulary items {len(lot_config_encoder.categories_[0])}\n")

lot_config_transformed = lot_config_encoder.transform(lot_config_list_of_lists)

lot_config_transformed = lot_config_transformed.toarray()

print(lot_config_transformed)

Unique vocabulary items 5

[[0. 0. 0. 0. 1.]
 [0. 0. 1. 0. 0.]
 [0. 0. 0. 0. 1.]
 ...
 [0. 0. 0. 0. 1.]
 [0. 0. 0. 0. 1.]
 [0. 0. 0. 0. 1.]]


In [22]:
housing_features_df["LotConfig"].head()

0    Inside
1       FR2
2    Inside
3    Corner
4       FR2
Name: LotConfig, dtype: object

In [23]:
type(lot_config_transformed)

numpy.ndarray

We have the one hot encodings for lot configurations so we should proceed to concatenate them to the original dataset.

In [24]:
lot_config_transformed_df = pd.DataFrame(lot_config_transformed)
lot_config_transformed_df.head()


0    1    2    3    4
0  0.0  0.0  0.0  0.0  1.0
1  0.0  0.0  1.0  0.0  0.0
2  0.0  0.0  0.0  0.0  1.0
3  1.0  0.0  0.0  0.0  0.0
4  0.0  0.0  1.0  0.0  0.0

I'm not sure if we replaced the labels of 0, 1, 2, 3, and 4 in the top row from the lab but it seems like something that should be replaced.

In [25]:
column_names = list(lot_config_transformed_df.columns.values)
print(column_names)

[0, 1, 2, 3, 4]


I originally wanted to match up the lot configurations but couldn't decide what column 1 and 3 were (either FR1 or Cul-de-sac).

In [26]:
temp = list(lot_config_transformed_df[1])
temp.index(1.0)

16

This tells me the first occurence of that certain lot configuration which occurs at 16. We can find the actual configuration now. 

In [27]:
lot_config_transformed_df[15:17]

0    1    2    3    4
15  1.0  0.0  0.0  0.0  0.0
16  0.0  1.0  0.0  0.0  0.0

In [28]:
housing_features_df["LotConfig"][15:17]

15     Corner
16    CulDSac
Name: LotConfig, dtype: object

From here, we can determine that column 3 should be renamed to FR3 since thats the last option. 

In [29]:
lot_config_transformed_df.rename(columns = {0:'Corner', 1:'CulDSac', 2:'FR2', 3:'FR3', 4: 'Inside'}, inplace = True)

In [30]:
lot_config_transformed_df.head()

Corner  CulDSac  FR2  FR3  Inside
0     0.0      0.0  0.0  0.0     1.0
1     0.0      0.0  1.0  0.0     0.0
2     0.0      0.0  0.0  0.0     1.0
3     1.0      0.0  0.0  0.0     0.0
4     0.0      0.0  1.0  0.0     0.0

Just to check if they align...

In [31]:
print(lot_config_transformed_df.shape)
print(housing_features_df.shape)

(1460, 5)
(1460, 8)


We can see that lot_config_transformed_df dataframe has 5 columns correpsonding to the 5 different lot configurations while housing_features_df has 8 columns corresponding to the different features that we picked out including lot configurations. Both of them have 1460 rows so we can now proceed to concatenating them.  

In [32]:
housing_features_df.reset_index(drop=True, inplace=True)
lot_config_transformed_df.reset_index(drop=True, inplace=True)

housing_features_transformed_df = pd.concat([housing_features_df, lot_config_transformed_df], axis=1)
housing_features_transformed_df.shape

(1460, 13)

In [33]:
housing_features_transformed_df.head()

LotShape LotConfig Neighborhood  OverallQual  OverallCond  YearRemodAdd  \
0      Reg    Inside      CollgCr            7            5          2003   
1      Reg       FR2      Veenker            6            8          1976   
2      IR1    Inside      CollgCr            7            5          2002   
3      IR1    Corner      Crawfor            7            5          1970   
4      IR1       FR2      NoRidge            8            5          2000   

   GrLivArea  MoSold  Corner  CulDSac  FR2  FR3  Inside  
0       1710       2     0.0      0.0  0.0  0.0     1.0  
1       1262       5     0.0      0.0  1.0  0.0     0.0  
2       1786       9     0.0      0.0  0.0  0.0     1.0  
3       1717       2     1.0      0.0  0.0  0.0     0.0  
4       2198      12     0.0      0.0  1.0  0.0     0.0

In [34]:
housing_features_transformed_df.pop('LotConfig')

0       Inside
1          FR2
2       Inside
3       Corner
4          FR2
         ...  
1455    Inside
1456    Inside
1457    Inside
1458    Inside
1459    Inside
Name: LotConfig, Length: 1460, dtype: object

In [35]:
housing_features_transformed_df.head()

LotShape Neighborhood  OverallQual  OverallCond  YearRemodAdd  GrLivArea  \
0      Reg      CollgCr            7            5          2003       1710   
1      Reg      Veenker            6            8          1976       1262   
2      IR1      CollgCr            7            5          2002       1786   
3      IR1      Crawfor            7            5          1970       1717   
4      IR1      NoRidge            8            5          2000       2198   

   MoSold  Corner  CulDSac  FR2  FR3  Inside  
0       2     0.0      0.0  0.0  0.0     1.0  
1       5     0.0      0.0  1.0  0.0     0.0  
2       9     0.0      0.0  0.0  0.0     1.0  
3       2     1.0      0.0  0.0  0.0     0.0  
4      12     0.0      0.0  1.0  0.0     0.0

We still have one more categorical-nominal feature left to transform to one-hot encodings, Neighborhood. 

### Same one-hot encoding process is replicated for the categorical-nominal feature Neighborhood

In [36]:
housing_features_df["Neighborhood"].unique()

array(['CollgCr', 'Veenker', 'Crawfor', 'NoRidge', 'Mitchel', 'Somerst',
       'NWAmes', 'OldTown', 'BrkSide', 'Sawyer', 'NridgHt', 'NAmes',
       'SawyerW', 'IDOTRR', 'MeadowV', 'Edwards', 'Timber', 'Gilbert',
       'StoneBr', 'ClearCr', 'NPkVill', 'Blmngtn', 'BrDale', 'SWISU',
       'Blueste'], dtype=object)

In [37]:
neighborhood_list = housing_features_df["Neighborhood"].to_list()
print(neighborhood_list)

['CollgCr', 'Veenker', 'CollgCr', 'Crawfor', 'NoRidge', 'Mitchel', 'Somerst', 'NWAmes', 'OldTown', 'BrkSide', 'Sawyer', 'NridgHt', 'Sawyer', 'CollgCr', 'NAmes', 'BrkSide', 'NAmes', 'Sawyer', 'SawyerW', 'NAmes', 'NridgHt', 'IDOTRR', 'CollgCr', 'MeadowV', 'Sawyer', 'NridgHt', 'NAmes', 'NridgHt', 'NAmes', 'BrkSide', 'IDOTRR', 'Sawyer', 'CollgCr', 'NAmes', 'NridgHt', 'NridgHt', 'CollgCr', 'NAmes', 'NAmes', 'Edwards', 'NAmes', 'Timber', 'SawyerW', 'CollgCr', 'NAmes', 'NridgHt', 'Mitchel', 'Somerst', 'OldTown', 'Sawyer', 'Gilbert', 'BrkSide', 'IDOTRR', 'Veenker', 'NAmes', 'NAmes', 'Somerst', 'CollgCr', 'StoneBr', 'CollgCr', 'SawyerW', 'IDOTRR', 'NridgHt', 'OldTown', 'CollgCr', 'NridgHt', 'NAmes', 'CollgCr', 'OldTown', 'ClearCr', 'NAmes', 'Mitchel', 'Gilbert', 'NAmes', 'OldTown', 'MeadowV', 'NAmes', 'BrkSide', 'Sawyer', 'OldTown', 'NAmes', 'Mitchel', 'Somerst', 'NAmes', 'Gilbert', 'NoRidge', 'Gilbert', 'Somerst', 'IDOTRR', 'CollgCr', 'NAmes', 'NAmes', 'Crawfor', 'OldTown', 'CollgCr', 'Gilbert

In [38]:
neighborhood_list_of_lists = []

for i in neighborhood_list: 
    neighborhood_list_of_lists.append([i])

print(neighborhood_list_of_lists[:5])

[['CollgCr'], ['Veenker'], ['CollgCr'], ['Crawfor'], ['NoRidge']]


In [39]:
neighborhood_encoder = OneHotEncoder()

neighborhood_encoder.fit(neighborhood_list_of_lists)

print(f"Unique vocabulary items {len(neighborhood_encoder.categories_[0])}\n")

neighborhood_transformed = neighborhood_encoder.transform(neighborhood_list_of_lists)

neighborhood_transformed = neighborhood_transformed.toarray()

print(neighborhood_transformed)

Unique vocabulary items 25

[[0. 0. 0. ... 0. 0. 0.]
 [0. 0. 0. ... 0. 0. 1.]
 [0. 0. 0. ... 0. 0. 0.]
 ...
 [0. 0. 0. ... 0. 0. 0.]
 [0. 0. 0. ... 0. 0. 0.]
 [0. 0. 0. ... 0. 0. 0.]]


In [40]:
neighborhood_transformed_df = pd.DataFrame(neighborhood_transformed)
neighborhood_transformed_df.head()

0    1    2    3    4    5    6    7    8    9   ...   15   16   17   18  \
0  0.0  0.0  0.0  0.0  0.0  1.0  0.0  0.0  0.0  0.0  ...  0.0  0.0  0.0  0.0   
1  0.0  0.0  0.0  0.0  0.0  0.0  0.0  0.0  0.0  0.0  ...  0.0  0.0  0.0  0.0   
2  0.0  0.0  0.0  0.0  0.0  1.0  0.0  0.0  0.0  0.0  ...  0.0  0.0  0.0  0.0   
3  0.0  0.0  0.0  0.0  0.0  0.0  1.0  0.0  0.0  0.0  ...  0.0  0.0  0.0  0.0   
4  0.0  0.0  0.0  0.0  0.0  0.0  0.0  0.0  0.0  0.0  ...  1.0  0.0  0.0  0.0   

    19   20   21   22   23   24  
0  0.0  0.0  0.0  0.0  0.0  0.0  
1  0.0  0.0  0.0  0.0  0.0  1.0  
2  0.0  0.0  0.0  0.0  0.0  0.0  
3  0.0  0.0  0.0  0.0  0.0  0.0  
4  0.0  0.0  0.0  0.0  0.0  0.0  

[5 rows x 25 columns]

There is probably a better way to rename the columns from one-hot encoding but this is the method thats available to me as of now. 

In [41]:
temp = list(neighborhood_transformed_df[0])
temp.index(1.0)

219

In [42]:
housing_features_df["Neighborhood"][219:220]

219    Blmngtn
Name: Neighborhood, dtype: object

In [43]:
temp = list(neighborhood_transformed_df[1])
temp.index(1.0)

599

In [44]:
housing_features_df["Neighborhood"][599]

'Blueste'

In [45]:
temp_list = []
for i in range(len(neighborhood_transformed_df.columns)):
    temp = list(neighborhood_transformed_df[i])
    temp_list.append(temp.index(1.0))

In [46]:
housing_features_df["Neighborhood"].head()

0    CollgCr
1    Veenker
2    CollgCr
3    Crawfor
4    NoRidge
Name: Neighborhood, dtype: object

In [47]:
neighborhood_name_list = []
for i in temp_list:
    neighborhood_name_list.append(housing_features_df["Neighborhood"][i])
print(neighborhood_name_list)

['Blmngtn', 'Blueste', 'BrDale', 'BrkSide', 'ClearCr', 'CollgCr', 'Crawfor', 'Edwards', 'Gilbert', 'IDOTRR', 'MeadowV', 'Mitchel', 'NAmes', 'NPkVill', 'NWAmes', 'NoRidge', 'NridgHt', 'OldTown', 'SWISU', 'Sawyer', 'SawyerW', 'Somerst', 'StoneBr', 'Timber', 'Veenker']


In [48]:
neighborhood_transformed_df.columns = neighborhood_name_list

In [49]:
neighborhood_transformed_df.head()

Blmngtn  Blueste  BrDale  BrkSide  ClearCr  CollgCr  Crawfor  Edwards  \
0      0.0      0.0     0.0      0.0      0.0      1.0      0.0      0.0   
1      0.0      0.0     0.0      0.0      0.0      0.0      0.0      0.0   
2      0.0      0.0     0.0      0.0      0.0      1.0      0.0      0.0   
3      0.0      0.0     0.0      0.0      0.0      0.0      1.0      0.0   
4      0.0      0.0     0.0      0.0      0.0      0.0      0.0      0.0   

   Gilbert  IDOTRR  ...  NoRidge  NridgHt  OldTown  SWISU  Sawyer  SawyerW  \
0      0.0     0.0  ...      0.0      0.0      0.0    0.0     0.0      0.0   
1      0.0     0.0  ...      0.0      0.0      0.0    0.0     0.0      0.0   
2      0.0     0.0  ...      0.0      0.0      0.0    0.0     0.0      0.0   
3      0.0     0.0  ...      0.0      0.0      0.0    0.0     0.0      0.0   
4      0.0     0.0  ...      1.0      0.0      0.0    0.0     0.0      0.0   

   Somerst  StoneBr  Timber  Veenker  
0      0.0      0.0     0.0      0.0  
1      0.0      0.0     0.0      1.0  
2      0.0      0.0     0.0      0.0  
3      0.0      0.0     0.0      0.0  
4      0.0      0.0     0.0      0.0  

[5 rows x 25 columns]

After labeling the columns with the correct neighborhood names, I can now concatenate this with the housing_features_transformed_df. To check...

In [50]:
print(neighborhood_transformed_df.shape)
print(housing_features_transformed_df.shape)

(1460, 25)
(1460, 12)


Since the rows align, we should be able to concatenate the one-hot encodings to the dataframe. 

In [51]:
housing_features_transformed_df.reset_index(drop=True, inplace=True)
neighborhood_transformed_df.reset_index(drop=True, inplace=True)

housing_features_final_df = pd.concat([housing_features_transformed_df, neighborhood_transformed_df], axis=1)
housing_features_final_df.shape

(1460, 37)

In [52]:
housing_features_final_df.head()

LotShape Neighborhood  OverallQual  OverallCond  YearRemodAdd  GrLivArea  \
0      Reg      CollgCr            7            5          2003       1710   
1      Reg      Veenker            6            8          1976       1262   
2      IR1      CollgCr            7            5          2002       1786   
3      IR1      Crawfor            7            5          1970       1717   
4      IR1      NoRidge            8            5          2000       2198   

   MoSold  Corner  CulDSac  FR2  ...  NoRidge  NridgHt  OldTown  SWISU  \
0       2     0.0      0.0  0.0  ...      0.0      0.0      0.0    0.0   
1       5     0.0      0.0  1.0  ...      0.0      0.0      0.0    0.0   
2       9     0.0      0.0  0.0  ...      0.0      0.0      0.0    0.0   
3       2     1.0      0.0  0.0  ...      0.0      0.0      0.0    0.0   
4      12     0.0      0.0  1.0  ...      1.0      0.0      0.0    0.0   

   Sawyer  SawyerW  Somerst  StoneBr  Timber  Veenker  
0     0.0      0.0      0.0      0.0     0.0      0.0  
1     0.0      0.0      0.0      0.0     0.0      1.0  
2     0.0      0.0      0.0      0.0     0.0      0.0  
3     0.0      0.0      0.0      0.0     0.0      0.0  
4     0.0      0.0      0.0      0.0     0.0      0.0  

[5 rows x 37 columns]

In [53]:
housing_features_final_df.pop("Neighborhood")

0       CollgCr
1       Veenker
2       CollgCr
3       Crawfor
4       NoRidge
         ...   
1455    Gilbert
1456     NWAmes
1457    Crawfor
1458      NAmes
1459    Edwards
Name: Neighborhood, Length: 1460, dtype: object

In [54]:
housing_features_final_df.head()

LotShape  OverallQual  OverallCond  YearRemodAdd  GrLivArea  MoSold  Corner  \
0      Reg            7            5          2003       1710       2     0.0   
1      Reg            6            8          1976       1262       5     0.0   
2      IR1            7            5          2002       1786       9     0.0   
3      IR1            7            5          1970       1717       2     1.0   
4      IR1            8            5          2000       2198      12     0.0   

   CulDSac  FR2  FR3  ...  NoRidge  NridgHt  OldTown  SWISU  Sawyer  SawyerW  \
0      0.0  0.0  0.0  ...      0.0      0.0      0.0    0.0     0.0      0.0   
1      0.0  1.0  0.0  ...      0.0      0.0      0.0    0.0     0.0      0.0   
2      0.0  0.0  0.0  ...      0.0      0.0      0.0    0.0     0.0      0.0   
3      0.0  0.0  0.0  ...      0.0      0.0      0.0    0.0     0.0      0.0   
4      0.0  1.0  0.0  ...      1.0      0.0      0.0    0.0     0.0      0.0   

   Somerst  StoneBr  Timber  Veenker  
0      0.0      0.0     0.0      0.0  
1      0.0      0.0     0.0      1.0  
2      0.0      0.0     0.0      0.0  
3      0.0      0.0     0.0      0.0  
4      0.0      0.0     0.0      0.0  

[5 rows x 36 columns]

This is the final dataframe that we are left with. The categorical-nominal features, **LotConfig and Neighborhood**, are both transformed into one-hot encodings. LotConfig (5) and Neighborhood (25) leaving 6 of the original columns unchanged. 

### e) [Extra Credit] Feature Scaling

This part asks us to scale the numerical features of **housing_features_final_df** using MinMax Scaler from sklearn.preprocessing. The numerical features of this dataframe would be **YearRemodAdd and GrLivArea**. 

In [57]:
from sklearn.preprocessing import MinMaxScaler
scaler = MinMaxScaler()
housing_features_final_df[["YearRemodAdd", "GrLivArea"]] = scaler.fit_transform(housing_features_final_df[["YearRemodAdd", "GrLivArea"]])
housing_features_final_df.head()

LotShape  OverallQual  OverallCond  YearRemodAdd  GrLivArea  MoSold  Corner  \
0      Reg            7            5      0.883333   0.259231       2     0.0   
1      Reg            6            8      0.433333   0.174830       5     0.0   
2      IR1            7            5      0.866667   0.273549       9     0.0   
3      IR1            7            5      0.333333   0.260550       2     1.0   
4      IR1            8            5      0.833333   0.351168      12     0.0   

   CulDSac  FR2  FR3  ...  NoRidge  NridgHt  OldTown  SWISU  Sawyer  SawyerW  \
0      0.0  0.0  0.0  ...      0.0      0.0      0.0    0.0     0.0      0.0   
1      0.0  1.0  0.0  ...      0.0      0.0      0.0    0.0     0.0      0.0   
2      0.0  0.0  0.0  ...      0.0      0.0      0.0    0.0     0.0      0.0   
3      0.0  0.0  0.0  ...      0.0      0.0      0.0    0.0     0.0      0.0   
4      0.0  1.0  0.0  ...      1.0      0.0      0.0    0.0     0.0      0.0   

   Somerst  StoneBr  Timber  Veenker  
0      0.0      0.0     0.0      0.0  
1      0.0      0.0     0.0      1.0  
2      0.0      0.0     0.0      0.0  
3      0.0      0.0     0.0      0.0  
4      0.0      0.0     0.0      0.0  

[5 rows x 36 columns]

### Edits to the Original Assignment - LotShape is considered a categorical-nominal variable now

In [58]:
housing_features_final_df['LotShape'].unique()

array(['Reg', 'IR1', 'IR2', 'IR3'], dtype=object)

In [60]:
lot_shape_list = housing_features_final_df['LotShape'].tolist()
print(lot_shape_list)

['Reg', 'Reg', 'IR1', 'IR1', 'IR1', 'IR1', 'Reg', 'IR1', 'Reg', 'Reg', 'Reg', 'IR1', 'IR2', 'IR1', 'IR1', 'Reg', 'IR1', 'Reg', 'Reg', 'Reg', 'IR1', 'Reg', 'Reg', 'Reg', 'IR1', 'Reg', 'Reg', 'Reg', 'IR1', 'IR1', 'Reg', 'IR1', 'Reg', 'IR1', 'Reg', 'Reg', 'Reg', 'Reg', 'Reg', 'Reg', 'Reg', 'Reg', 'IR1', 'IR1', 'Reg', 'Reg', 'IR1', 'Reg', 'Reg', 'Reg', 'IR2', 'Reg', 'IR2', 'IR1', 'Reg', 'IR1', 'Reg', 'IR1', 'IR2', 'Reg', 'Reg', 'Reg', 'IR1', 'IR1', 'Reg', 'Reg', 'Reg', 'IR1', 'Reg', 'Reg', 'IR1', 'Reg', 'IR1', 'Reg', 'Reg', 'Reg', 'IR1', 'Reg', 'Reg', 'Reg', 'Reg', 'Reg', 'Reg', 'IR1', 'IR1', 'Reg', 'IR2', 'Reg', 'IR1', 'Reg', 'Reg', 'Reg', 'IR1', 'Reg', 'IR1', 'IR2', 'IR1', 'Reg', 'Reg', 'IR1', 'IR1', 'Reg', 'Reg', 'IR1', 'Reg', 'Reg', 'Reg', 'Reg', 'Reg', 'IR1', 'Reg', 'IR1', 'Reg', 'Reg', 'IR1', 'Reg', 'Reg', 'Reg', 'Reg', 'Reg', 'IR1', 'Reg', 'Reg', 'Reg', 'IR1', 'Reg', 'IR1', 'IR1', 'Reg', 'Reg', 'Reg', 'IR1', 'Reg', 'IR1', 'IR1', 'Reg', 'IR1', 'Reg', 'IR1', 'IR1', 'Reg', 'Reg', 'Reg'

In [62]:
lot_shape_list_of_lists = []

for i in lot_shape_list: 
    lot_shape_list_of_lists.append([i])

print(lot_shape_list_of_lists[:5])  

[['Reg'], ['Reg'], ['IR1'], ['IR1'], ['IR1']]


In [63]:
shape_encoder = OneHotEncoder()

shape_encoder.fit(lot_shape_list_of_lists)

print(f"Unique vocabulary items {len(shape_encoder.categories_[0])}\n")

shape_transformed = shape_encoder.transform(lot_shape_list_of_lists)

shape_transformed = shape_transformed.toarray()

print(shape_transformed)

Unique vocabulary items 4

[[0. 0. 0. 1.]
 [0. 0. 0. 1.]
 [1. 0. 0. 0.]
 ...
 [0. 0. 0. 1.]
 [0. 0. 0. 1.]
 [0. 0. 0. 1.]]


In [66]:
shape_transformed_df = pd.DataFrame(shape_transformed)
print(shape_transformed_df)

        0    1    2    3
0     0.0  0.0  0.0  1.0
1     0.0  0.0  0.0  1.0
2     1.0  0.0  0.0  0.0
3     1.0  0.0  0.0  0.0
4     1.0  0.0  0.0  0.0
...   ...  ...  ...  ...
1455  0.0  0.0  0.0  1.0
1456  0.0  0.0  0.0  1.0
1457  0.0  0.0  0.0  1.0
1458  0.0  0.0  0.0  1.0
1459  0.0  0.0  0.0  1.0

[1460 rows x 4 columns]


In [70]:
temp = list(shape_transformed_df[1])
temp.index(1.0)

12

In [71]:
housing_features_final_df[12:13]

LotShape  OverallQual  OverallCond  YearRemodAdd  GrLivArea  MoSold  \
12      IR2            5            6           0.2   0.108892       9   

    Corner  CulDSac  FR2  FR3  ...  NoRidge  NridgHt  OldTown  SWISU  Sawyer  \
12     0.0      0.0  0.0  0.0  ...      0.0      0.0      0.0    0.0     1.0   

    SawyerW  Somerst  StoneBr  Timber  Veenker  
12      0.0      0.0      0.0     0.0      0.0  

[1 rows x 36 columns]

In [72]:
shape_transformed_df.rename(columns = {0:'IR1', 1:'IR2', 2:'IR3', 3:'Reg'}, inplace = True)

In [73]:
shape_transformed_df.head()

IR1  IR2  IR3  Reg
0  0.0  0.0  0.0  1.0
1  0.0  0.0  0.0  1.0
2  1.0  0.0  0.0  0.0
3  1.0  0.0  0.0  0.0
4  1.0  0.0  0.0  0.0

In [77]:
print(shape_transformed_df.shape)
print(housing_features_final_df.shape)

(1460, 4)
(1460, 36)


In [78]:
housing_features_final_df.reset_index(drop=True, inplace=True)
shape_transformed_df.reset_index(drop=True, inplace=True)

official_features_df = pd.concat([housing_features_final_df, shape_transformed_df], axis=1)
official_features_df.shape

(1460, 40)

In [79]:
official_features_df.head()

LotShape  OverallQual  OverallCond  YearRemodAdd  GrLivArea  MoSold  Corner  \
0      Reg            7            5      0.883333   0.259231       2     0.0   
1      Reg            6            8      0.433333   0.174830       5     0.0   
2      IR1            7            5      0.866667   0.273549       9     0.0   
3      IR1            7            5      0.333333   0.260550       2     1.0   
4      IR1            8            5      0.833333   0.351168      12     0.0   

   CulDSac  FR2  FR3  ...  Sawyer  SawyerW  Somerst  StoneBr  Timber  Veenker  \
0      0.0  0.0  0.0  ...     0.0      0.0      0.0      0.0     0.0      0.0   
1      0.0  1.0  0.0  ...     0.0      0.0      0.0      0.0     0.0      1.0   
2      0.0  0.0  0.0  ...     0.0      0.0      0.0      0.0     0.0      0.0   
3      0.0  0.0  0.0  ...     0.0      0.0      0.0      0.0     0.0      0.0   
4      0.0  1.0  0.0  ...     0.0      0.0      0.0      0.0     0.0      0.0   

   IR1  IR2  IR3  Reg  
0  0.0  0.0  0.0  1.0  
1  0.0  0.0  0.0  1.0  
2  1.0  0.0  0.0  0.0  
3  1.0  0.0  0.0  0.0  
4  1.0  0.0  0.0  0.0  

[5 rows x 40 columns]

In [80]:
official_features_df.pop('LotShape')

0       Reg
1       Reg
2       IR1
3       IR1
4       IR1
       ... 
1455    Reg
1456    Reg
1457    Reg
1458    Reg
1459    Reg
Name: LotShape, Length: 1460, dtype: object

In [81]:
official_features_df.head()

OverallQual  OverallCond  YearRemodAdd  GrLivArea  MoSold  Corner  CulDSac  \
0            7            5      0.883333   0.259231       2     0.0      0.0   
1            6            8      0.433333   0.174830       5     0.0      0.0   
2            7            5      0.866667   0.273549       9     0.0      0.0   
3            7            5      0.333333   0.260550       2     1.0      0.0   
4            8            5      0.833333   0.351168      12     0.0      0.0   

   FR2  FR3  Inside  ...  Sawyer  SawyerW  Somerst  StoneBr  Timber  Veenker  \
0  0.0  0.0     1.0  ...     0.0      0.0      0.0      0.0     0.0      0.0   
1  1.0  0.0     0.0  ...     0.0      0.0      0.0      0.0     0.0      1.0   
2  0.0  0.0     1.0  ...     0.0      0.0      0.0      0.0     0.0      0.0   
3  0.0  0.0     0.0  ...     0.0      0.0      0.0      0.0     0.0      0.0   
4  1.0  0.0     0.0  ...     0.0      0.0      0.0      0.0     0.0      0.0   

   IR1  IR2  IR3  Reg  
0  0.0  0.0  0.0  1.0  
1  0.0  0.0  0.0  1.0  
2  1.0  0.0  0.0  0.0  
3  1.0  0.0  0.0  0.0  
4  1.0  0.0  0.0  0.0  

[5 rows x 39 columns]